### PR TITLE
bpo-46032: Check types in singledispatch's register() at declaration time

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
The registry() method of functools.singledispatch() functions checks now
the first argument or the first parameter annotation and raises a TypeError if it is
not supported. Previously unsupported "types" were ignored (e.g. typing.List[int])
or caused an error at calling time (e.g. list[int]).

https://bugs.python.org/issue46032

### CODE REVIEW

**Review by uriyyo — APPROVED** [review]
LGTM

File: Lib/functools.py [hunk 1]
```diff
@@ -867,7 +878,8 @@ def register(cls, func=None):
             # only import typing if annotation parsing is necessary
             from typing import get_type_hints
             argname, cls = next(iter(get_type_hints(func).items()))
-            if not isinstance(cls, type) and not _is_valid_union_type(cls):
+            if not ((isinstance(cls, type) and not isinstance(cls, GenericAlias))
```
Comment (AlexWaygood) [inline Lib/functools.py]: This is repeated logic from lines 858-859. Could this be factored out into an `_is_valid_dispatch_type` function? I think that would make it easier to modify if we decide to support other types in `singledispatch` in the future.

Comment (serhiy-storchaka) [reply Lib/functools.py]: Good idea. Done.

**Review by AlexWaygood — APPROVED** [review]
LGTM

**Pull request merged**